File: parser.js
const fs = require('fs')
const puppeteer = require('puppeteer')
const pMapSeries = require('p-map-series')
const { flatten } = require('lodash')
const BASE_URL = 'http://elezionistorico.interno.gov.it/index.php?tpel=C'

const readDates = () => {
  const dateElements = document.querySelectorAll('#collapseTwo .sezione_ristretta li a')
  const dateHrefs = Array.apply(undefined, dateElements).map(n => n.href)
  return dateHrefs
}

const readAreas = () => {
  const dateElements = document.querySelectorAll('#collapseThree .sezione li a')
  const dateHrefs = Array.apply(undefined, dateElements).map(n => n.href)
  return dateHrefs
}

const readCircumscription = () => {
  const dateElements = document.querySelectorAll('#collapseFour .sezione li a')
  const dateHrefs = Array.apply(undefined, dateElements).map(n => n.href)
  return dateHrefs
}

const readProvince = () => {
  const dateElements = document.querySelectorAll('#collapseFive .sezione li a')
  const dateHrefs = Array.apply(undefined, dateElements).map(n => n.href)
  return dateHrefs
}

const readCity = () => {
  const dateElements = document.querySelectorAll('#collapseSix .sezione li a')
  const dateHrefs = Array.apply(undefined, dateElements).map(n => n.href)
  return dateHrefs
}

function Parser() {
  const self = this
  self.dateHrefs = []
  self.areaHrefs = []
  self.circumscriptionHrefs = []
  self.provinceHrefs = []
  self.cityHrefs = []

  self.start = async () => {
    self.browser = await puppeteer.launch({
      headless: true,
    })
  }

  self.newPage = async () => {
    const page = await self.browser.newPage()
    return page
  }

  self.getAllPages = async () => {
    self.dateHrefs = await self.getDates()
    console.log(`${self.dateHrefs.length} dates`)
    const nestedAreas = await pMapSeries(self.dateHrefs, self.getAreas)
    self.areaHrefs = flatten(nestedAreas)
    console.log(`${self.areaHrefs.length} areas`)
    const nestedCircumscriptions = await pMapSeries(self.areaHrefs, self.getCircumscriptions)
    self.circumscriptionHrefs = flatten(nestedCircumscriptions)
    console.log(`${self.circumscriptionHrefs.length} circums`)
    const nestedProvinces = await pMapSeries(self.circumscriptionHrefs, self.getProvinces)
    self.provinceHrefs = flatten(nestedProvinces)
    console.log(`${self.provinceHrefs.length} provinces`)
    await pMapSeries(self.provinceHrefs, self.getCities)
    // const nestedCities = await pMapSeries(self.provinceHrefs, self.getCities)
    // self.cityHrefs = flatten(nestedCities)
  }

  self.getDates = async () => {
    const page = await self.newPage()
    await page.goto(BASE_URL, { waitUntil: 'load' })
    const hrefs = await page.evaluate(readDates)
    await page.close()
    return hrefs
  }

  self.getAreas = async url => {
    const page = await self.newPage()
    await page.goto(url, { waitUntil: 'load' })
    const hrefs = await page.evaluate(readAreas)
    await page.close()
    return hrefs
  }

  self.getCircumscriptions = async url => {
    const page = await self.newPage()
    await page.goto(url, { waitUntil: 'load' })
    const hrefs = await page.evaluate(readCircumscription)
    await page.close()
    return hrefs
  }

  self.getProvinces = async url => {
    const page = await self.newPage()
    await page.goto(url, { waitUntil: 'load' })
    const hrefs = await page.evaluate(readProvince)
    await page.close()
    return hrefs
  }

  self.getCities = async url => {
    const page = await self.newPage()
    await page.goto(url, { waitUntil: 'load' })
    const hrefs = await page.evaluate(readCity)
    await page.close()
    hrefs.forEach(href => {
      fs.appendFileSync('./all-hrefs.txt', `${href}\n`)
    })
  }
}

const start = async () => {
  const parser = new Parser()
  await parser.start()
  await parser.getAllPages()
  console.log('done')
  process.exit()
}

start()

File: parser.py
import scrapy
from scrapy.http import Request
from pdb import set_trace as shtap
from random import choice

BASE_URL = 'http://elezionistorico.interno.gov.it'
FIRST_PATH = '/index.php?tpel=C'

class ElectionSpider(scrapy.Spider):
  name = 'electionSpider'
  start_urls = [BASE_URL + FIRST_PATH]
  download_delay = 2

  def parse(self, response):
    for date in response.css('#collapseTwo .sezione_ristretta li a'):
      path = date.css('a::attr(href)').extract_first()
      request = Request(url=BASE_URL + path, callback=self.parse_date)
      yield request

  def parse_date(self, response):
    for area in response.css('#collapseThree .sezione li a'):
      path = area.css('a::attr(href)').extract_first()
      request = Request(BASE_URL + path, callback=self.parse_area)
      yield request

  def parse_area(self, response):
    for area in response.css('#collapseFour .sezione li a'):
      path = area.css('a::attr(href)').extract_first()
      request = Request(BASE_URL + path, callback=self.parse_circumscription)
      yield request

  def parse_circumscription(self, response):
    for area in response.css('#collapseFive .sezione li a'):
      path = area.css('a::attr(href)').extract_first()
      request = Request(BASE_URL + path, callback=self.parse_province)
      yield request

  def parse_province(self, response):
    for area in response.css('#collapseSix .sezione li a'):
      path = area.css('a::attr(href)').extract_first()
      final_path = (BASE_URL + path + '&ord=4').replace('index', 'estrazione')
      yield { "href": final_path }
